### DISCUSSION TITLE
Slicing support for multi-dimensional collections

### DISCUSSION THREAD

**Discussion by eiriktsarpalis** [dotnet/csharplang#7935] [Language Ideas]
In https://github.com/dotnet/csharplang/issues/7913#issuecomment-1936505446 I proposed adding support for multi-dimensional collection literals. This would be an important addition facilitating usage of the new tensor library that we'll be working on for .NET 9.

One feature request that dovetails with the above is support for multi-dimensional slicing.  In pytorch, it is possible to [create slices](https://pytorch.org/tensordict/tutorials/tensordict_slicing.html) using the following syntax:
```python
tensor[:, 2:, :3, 5:10] # gets a view of the sub-tensor scoped to the specified dimensions
```
Slices can also be assigned to as the following example illustrates:
```python
tensor[0, :] = torch.ones(3) # fill the first row of tensor with ones
```

Transcribing the above to C# slicing syntax might look as follows
```C#
Tensor<T> tensor = ...;
Tensor<T> view = tensor[.., 2.., ..3, 5..10];
tensor[0, ..] = Tensor<int>.Ones(3);
```

## Sketch Design

A strawman implementation of multi-dimensional slicing on collection types of fixed rank might look as follows
```C#
public class Matrix2D<T>
{
    public Matrix2D<T> Get_Slice(Range range1, Range range2);
    public void Set_Slice(Range range1, Range range2, Matrix2D<T> value);
}
```
Whereas for variable rank types such as tensors it might look as follows:
```C#
public class Tensor<T>
{
    public Tensor<T> Get_Slice(ReadOnlySpan<Range> ranges);
    public void Set_Slice(ReadOnlySpan<Range> ranges, Tensor<T> value);
}
```

Given the above, the previous expressions should desugar into
```C#
Tensor<T> view = tensor.Get_Slice([(..), 2.., (..3), 5..10]);
tensor.Set_Slice([0..1, (..)], ones);
```

Thoughts? Is this something we could pursue for the next C#?

## Bonus: The ellipsis operator

Python defines `...` as a [special constant](https://docs.python.org/3/library/constants.html#Ellipsis) that among other things can be passed as a parameter in indexer methods. The code
```python
tensor[..., 0]
```
can be used as a synonym to writing
```python
tensor[:, :, :, 0]
```
It would be nice if we could achieve something similar in C# slicers, however it should be noted that this cannot be implemented as syntactic sugar for the case of tensors -- the rank of the tensor cannot be known at compile time. Most likely it would require runtime representation (potentially as an augmentation of the `System.Range` type?) which the slicer can interpret accordingly.

**Comment by tannergooding** [comment]
This likely requires expansion to the `System.Index` and `System.Range` support to also include `long` and/or `nint` -- (ideally we could just support any numeric like type or type that follows a required shape)

**Reply by tannergooding** [reply]
Perhaps this could be `System.Index<T> where T : INumber<T>` or `System.Index<T> where T : IBinaryInteger<T>`

**Reply by NiklasGustafsson** [reply]
I agree that it would be very useful to extend System.Index to other integer types.

**Reply by NiklasGustafsson** [reply]
It's unrelated to indexing / slicing, but would float ranges also have value, perhaps?

**Reply by eiriktsarpalis** [reply]
Per @NiklasGustafsson's comment in https://github.com/dotnet/csharplang/discussions/7935#discussioncomment-8445587, if we wanted to support ellipses in the style of pytorch, we'd need to augment the existing `Range` type which may or may not be possible.

**Reply by NiklasGustafsson** [reply]
Or make it a special value, like `System.Ellipsis`

**Reply by eiriktsarpalis** [reply]
Making it a distinct type would mean that we couldn't accept it as an argument in the slicer methods.

**Reply by NiklasGustafsson** [reply]
Good point

**Reply by tannergooding** [reply]
> It's unrelated to indexing / slicing, but would float ranges also have value, perhaps?

IMO, no. Floating-point values start getting into a lot of complexities that can't be trivially handled or explained.

This includes needing to track a step amount so that `0.5..0.75` could "make sense" and explaining concepts like what happens for `0..33554432.0f`, given that once you hit `16777216` (2^24) integers aren't exactly representable anymore and thus `16777216 + 1` is `16777216`, not `16777217`. At that point, representable values are `2` apart; then `4` apart at `2^25`, `8 apart at `2^26`, and so on.

**Reply by NiklasGustafsson** [reply]
It wouldn't deal with the step issue, just be a way to represent [a,b) in a nicer way, but if System.Range has to have enumeration semantics, then it clearly isn't applicable.

**Reply by eiriktsarpalis** [reply]
`System.Range` was very deliberately designed as a type meant to represent a range of (non-negative) indices. I think we might need a separate type to represent arbitrary numeric intervals such as `[a, b)`. We might want to call it `Interval<T>`.

**Reply by tannergooding** [reply]
`System.Range` is specifically designed to coordinate with `System.Index` and has properties such as `All` or `GetOffsetAndLength`, which requires some interesting considerations when factoring in floating-point values and could potentially lead to issues.

So while it isn't strictly "needs enumeration", it isn't straightforward to support.

**Reply by NiklasGustafsson** [reply]
Some other languages conflate the two, but it makes sense given the design history of System.Range in .NET

Without syntactic sugar, intervals can just as well be represented by a tuple (a triple when you want to provide the step). 

Sorry about the tangent... now back to regular programming! :-)

**Comment by NiklasGustafsson** [comment]
Python also has a value denoted '...' which is a wildcard over dimensions and allows you to write code that is generic over the number of dimensions. 

It may be most useful for variable-rank types, but could also be used for something like:

```C#
tensor[...]  = Tensor<int>.Ones(3);
```

where it would overwrite the contents of 'tensor', rather than modify the variable 'tensor'

**Comment by CyrusNajmabadi** [comment]
Question.  Why is 

```c#
tensor[:, 2:, :3, 5:10] # gets a view of the sub-tensor scoped to the specified dimensions
```

that much better than:

```c#
tensor[.., 2.., ..3, 5..10];
```

Is it just to ape python?  That's not hugely compelling to me.  With the C# 13 work to support params collections, we should be able to have:

```c#
public Tensor<T> this(params ReadOnlySpan<Range> ranges);
```

so you would literally be able to write the latter and have it work.

**Reply by NiklasGustafsson** [reply]
I think that's what Eirik was suggesting.

**Reply by CyrusNajmabadi** [reply]
> I think that's what Eirik was suggesting.

Then this would need no language support.  It would already be possible in c# 13 with the params-span work.

**Reply by eiriktsarpalis** [reply]
Using indexer syntax? Or would it have to be a regular method?

**Reply by CyrusNajmabadi** [reply]
This:

```c#
tensor[.., 2.., ..3, 5..10];
```

Is legal today if you have an indexer that takes 4 range instances.

```c#
tensor[.., 2.., ..3, 5..10];
```

will also be legal in C#13 if you have an indxer which takes a `params ReadOnlySpan<Range>`.  

So having that specific syntax 'fall out' seems like it will already happen in C# 13.  :)  

Note: other requests (like ellipses `...` support) would still need lang work.

**Reply by eiriktsarpalis** [reply]
It seems that this API shape would work for us (using C# 13 syntax)

```C#
public class Tensor<T>
{
    public Tensor<T> this[params ReadOnlySpan<Range> ranges]
    {
        get => throw new NotImplementedException();
        set => throw new NotImplementedException();
    }
}
```
The only missing ingredient is supporting integer or `Index` coordinates. In principle it should be addressable using an implicit conversion from `int` or `Index` to `Range`.

**Reply by eiriktsarpalis** [reply]
One open question is how such an arrangement would interact in the presence of a regular indexer, from an overload resolution perspective:
```C#
public class Tensor<T>
{
    public T this[params ReadOnlySpan<Index> ranges]
    {
        get => throw new NotImplementedException();
        set => throw new NotImplementedException();
    }

    public Tensor<T> this[params ReadOnlySpan<Range> ranges]
    {
        get => throw new NotImplementedException();
        set => throw new NotImplementedException();
    }
}
```
In principle I think it should be ok, and the correct overload would be selected depending on the arguments being passed but I would need to test this in practice. I might trial this approach against multi-dimensional arrays in a local prototype.

**Reply by NiklasGustafsson** [reply]
Implicit conversions from integers to both Index and Range would make this ambiguous, I think. We don't have the shortest-chain-of-conversions rule that C++ has in C#, right?

**Reply by tannergooding** [reply]
> `public Tensor<T> this[params ReadOnlySpan<Range> ranges]`

I expect this may be the wrong signature. Other types do this via an API named `Slice` instead.

There's some nuance to be worked out across allocations, implicit costs, and consistency with other BCL/FDG concepts, etc

**Reply by eiriktsarpalis** [reply]
Existing `Slice` methods don't have the concept of a setter though, so some deviations might be necessary.

**Reply by tannergooding** [reply]
Yes, and that is somewhat "by design". C# doesn't have the concept of overriding `operator =()` like C/C++ does.

Given how .NET typically works, one wouldn't expect that this does `CopyTo`:
```csharp
Tensor<T> x = ...;
Tensor<T> y = ...;
x = y;
```

Instead, one would expect that `x` and `y` are now the same reference.

Thus, similarly one would expect that this does the same:
```csharp
Tensor<T> x = ...;
Tensor<T> y = ...;
x[..., ...] = y;
```

In order to do a copy operation, you would make it explicit, such as:
```csharp
Tensor<T> x = ...;
Tensor<T> y = ...;
y.CopyTo(x[..., ...]);
```

Which then also works as expected with value types, returned properties, etc. I am skeptical that API review would go for a new concept (where `=` suddenly means `CopyTo`) given what we already have and the limitations around the "new thing".

**Reply by NiklasGustafsson** [reply]
I don't see it that way -- it doesn't mean that `=` suddenly means `CopyTo`, it means that the `set` of `this[...]` gets called.

**Reply by tannergooding** [reply]
That's exactly the behavior you're describing here, is `CopyTo`.

The signature of `set` would literally be `instance void set_Item(ReadOnlySpan<Range> ranges, Tensor<T> value)` and that is how some languages would see it.

Such a signature then doesn't flow with .NET design guidelines or other existing types. For example, you cannot do `array[1..] = otherArray` (same goes for spans or any other built-in collection) and the compiler instead surfaces `error CS0131: The left-hand side of an assignment must be a variable, property or indexer`. It does this because there is a huge distinction between `indexing` something and `slicing` something. This then impacts what it then means to assign what a slice produces, what is intuitive when contrasted to the other APIs/operators/indexers exposed by the type, etc.

We have existing patterns for this and that work when correctly integrated to the language. We should not be trying to invent net new patterns which will bifurcate the ecosystem or otherwise lead users to pits of failure due to accidental bugs/etc. We should instead be in the mindset of "make what is correct for .NET". We can of course take inspiration from other ecosystems where that is appropriate and to help push the design in the right direction where questions come up, but we shouldn't be trying to deviate without good reason.

In this case, the existing pattern is an explicit `CopyTo` method. It makes it clear what the code is doing, it makes it clear the expense around what is being done. It avoids potential bugs when done in relation to value types. And so on.

**Reply by NiklasGustafsson** [reply]
Understood. 

There's a tension between design guidelines for existing .NET collection types and what numerical libraries outside the .NET ecosystem do. 

It's fair to say that 'we shall adhere to the existing .NET guidelines for collection types no matter what,' but I am of the opinion that it's worth considering whether those guidelines serve the scenarios.

`t.CopyTo(...)` isn't horrendous, but it's not as elegant as how other languages surface partial mutation of tensors.

**Reply by tannergooding** [reply]
Many other ecosystem don't support as many languages as .NET does and/or don't have the same level of care/concern around back/forward compat.

.NET is one of the few ecosystems where the same core that we shipped 20 years ago in .NET Framework 1.0 remains just as valid and good today as it was back then. We do make mistakes and there are places where we've opted to evolve things, but those are almost all considered extremely carefully and in the context of the broader ecosystem, where it's been, and where it's going.

What's "elegant" today can often lead to bugs, confusion, or errors 5 years down the road. So it's very important that we take in the broader context and understand what types of issues might crop up before doing something "new", even if its something that happens to work for another ecosystem today.

**Reply by tannergooding** [reply]
Consistency is also a major selling point for us. When we go and start supporting something new, we often want to ensure it can be integrated throughout the ecosystem cleanly and makes sense/works in most scenarios.

So, if we started saying that `x[1..] = y` (given `Tensor<T> x;` and `Tensor<T> y;` should work for `Tensor<T>`. Then you can expect that it needs to also work/make sense for `T[]`, `Span<T>`, `List<T>`, and other collection types.

**Reply by eiriktsarpalis** [reply]
@NiklasGustafsson correct if I'm wrong, but such forms are already possible in TorchSharp right?

**Reply by NiklasGustafsson** [reply]
Yes. 

TorchSharp was specifically designed to make code look as much like PyTorch code as possible, and since assigning to a slice is a common operation, it's supported. That includes assigning into a slice indexed by a Boolean tensor (a filter, in other words), which is far more useful in the target of an assignment than on the right-hand side.

You can do both:

```C#
t[0,..,..] = ...;
```
and

```C#
t[t < 5] = ...;
```

**Reply by eiriktsarpalis** [reply]
While I don't have a strong opinion on the matter, it might be worth considering that

* To my knowledge we have no precedent of assignable slice expressions, which should give us some flexibility wrt defining the appropriate convention.
* Conventions being followed around indexer implementations aren't universal. For example, it might be argued that a getter has dereferencing semantics and/or is idempotent, however that isn't the case with `Slice` methods in other collection types which have copy semantics. In other words
    ```C#
    List<int> list = [1, 2, 3];
    list[1..][0] = -1;
    list[1..][0]; // 2
    ```
    This would suggest to me that the semantics of slicing indexers in particular need to be considered to separately.
* I don't think pursuing consistency for the sake of consistency is necessarily beneficial. As mentioned, existing collection types (excluding span) use copy semantics for slicing, however doing the same thing for tensor types would be impractical.
* Finally, I want to echo @NiklasGustafsson's earlier point, it's worth considering whether the guidelines serve the scenarios. Slicing appears to be a core functionality of tensor libraries, and we would be offering this component either to data scientists already familiar with tensor libraries or people looking to port existing tensor code to .NET. In both cases maintaining a degree of mechanical sympathy seems like a good goal to me.

**Comment by NiklasGustafsson** [comment]
Note that this statement in Eirik's initial post:

```C#
tensor[0, ..] = Tensor<int>.Ones(3);
```
would require an implicit conversion from integer to System.Range, so that `a` is equivalent to `a..(a+1)`

**Reply by tannergooding** [reply]
Not quite sure I understand this?

`tensor[0, ..]` would effectively slice. It would end up producing a `TensorSpan` (for lack of a better name) where there is `1 row` and `all columns` (`C`). This could logically be represented as 2-dimensional `Cx1` or just 1-dimension `C`.

**Reply by NiklasGustafsson** [reply]
Okay, maybe I've misunderstood. 

For it to slice, I imagined that all the arguments have to be System.Range, which '0' isn't.

**Reply by tannergooding** [reply]
Ah, I see what you mean. Yes, given today's language support it would have to be `tensor[0..1, ..]` or similar to work.

To support just `a` (`0`) in this case, we'd need special language support for multiple dimensions or some allowed implicit conversion so that it works as selecting just the 1 specified

**Reply by CyrusNajmabadi** [reply]
or allow an int as the first arg?  But i don't know what hte expected args are?  Can you have such an int at any location, combinatorially, with all the other args?

**Reply by NiklasGustafsson** [reply]
Yes, it can appear anywhere in the list.

One could also declare an indexer taking a separate type, to which both System.Range and integers are converted implicitly. That's the approach we took in TorchSharp.

**Reply by CyrusNajmabadi** [reply]
> One could also declare an indexer taking a separate type, to which both System.Range and integers are converted implicitly. That's the approach we took in TorchSharp.

That actually seems... nice?

**Reply by NiklasGustafsson** [reply]
From TorchSharp:

```C#
public Tensor this[params long[] indices] { ... }

public Tensor this[params TensorIndex[] indices] { ... }
```

Where TensorIndex is the separate type... No special language support needed. If there are only integer indices, the first matches, otherwise the second.

**Reply by CyrusNajmabadi** [reply]
right.  note: you could still keep the second as an indexer to keep the simple `x[.......]` syntax

**Reply by CyrusNajmabadi** [reply]
And params span will allow the above to work more nicely without needing all the intermediary temp arrays.

**Reply by tannergooding** [reply]
We'd want to ensure such a thing integrated nicely with any range/index related language features and that it didn't cause overload resolution problems.

We may also want to consider if such a thing is tensor specific or reusable for any type of MD data (arrays, custom data types, etc)

**Reply by NiklasGustafsson** [reply]
But that's for tensors. Multi-dimensional arrays would need something with more tight language support, right? You'd want the compiler to check that it's got the right number of indices, for example.

**Reply by CyrusNajmabadi** [reply]
We're discussing md-arrays over in Collection-Exprs Next :)

**Reply by tannergooding** [reply]
Yes and ideally we don't have two separate types to support the general concept of multi-dimensional indexing/slicing.

One might get specialized language validation for MD arrays; but other types might not.

**Reply by NiklasGustafsson** [reply]
Yes, Eirik and I were having a conversation about collection expressions; slicing / ellipsis came up, which seemed orthogonal to that discussion, so Eirik started this.

**Reply by NiklasGustafsson** [reply]
BTW, the `TensorIndex` approach also addresses the ellipsis issue, since there can be a `TensorIndex.Ellipsis` singleton. Not as elegant and readable as '...', but good enough.
